### PR TITLE
Operations: callback not always called

### PR DESCRIPTION
**Pull request by rubenvp8510**
Issue #79: Operations: callback not always called

### CODE REVIEW

File: lib/hawkular/operations/operations_api.rb [hunk 1]
```diff
@@ -215,7 +215,10 @@ def check_pre_conditions(hash = {}, params = [], &callback)
       fail 'callback must have the perform method defined. include Hawkular::Operations' unless
           callback.nil? || callback.respond_to?('perform')
       params.each do |property|
-        fail "Hash property #{property} must be specified" if hash[property].nil?
+        if hash[property].nil?
+          error_message = "Hash property #{property} must be specified"
+          callback.nil? ? fail(error_message) : callback.perform(:failure, error_message)
```
Comment (josejulio) [inline lib/hawkular/operations/operations_api.rb]: I guess that it would be better to raise an [ArgumentError](http://apidock.com/ruby/ArgumentError) instead of the [RuntimeError of fail](http://ruby-doc.org/core-2.0.0/Kernel.html#method-i-fail)

File: spec/integration/operations_spec.rb [hunk 2]
```diff
@@ -86,6 +86,62 @@ module Hawkular::Operations::RSpec
...
+                                      })
+
+        noop = { operationName: 'noop' }
+        begin
+          client.invoke_generic_operation(noop)
+        rescue
+          puts 'We got an exception and this is good'
+        else
+          fail 'Should have failed with error:Hash property must be specified'
+        end
```
Comment (josejulio) [inline spec/integration/operations_spec.rb]: You could use something like:
expect { client.invoke_generic_operation(noop) }.to raise_error(ArgumentError)

Comment (jkremser) [inline spec/integration/operations_spec.rb]: +1 the "We got an exception and this is good" is polluting the output of rspec

File: lib/hawkular/operations/operations_api.rb [hunk 3]
```diff
@@ -215,7 +215,10 @@ def check_pre_conditions(hash = {}, params = [], &callback)
       fail 'callback must have the perform method defined. include Hawkular::Operations' unless
           callback.nil? || callback.respond_to?('perform')
       params.each do |property|
-        fail "Hash property #{property} must be specified" if hash[property].nil?
+        if hash[property].nil?
+          error_message = "Hash property #{property} must be specified"
+          callback.nil? ? fail(ArgumentError, error_message) : callback.perform(:failure, error_message)
```
Comment (abonas) [inline lib/hawkular/operations/operations_api.rb]: I might be missing something here, but here if callback is empty then it throws an ArgumentError. A few lines up on line 216, there is "fail .... **unless** callback.nil?" check that sounds a bit contradictory.

File: spec/integration/operations_spec.rb [hunk 4]
```diff
@@ -86,6 +86,56 @@ module Hawkular::Operations::RSpec
       end
     end
 
+    it 'should run into error callback because bad hash parameters' do
```
Comment (abonas) [inline spec/integration/operations_spec.rb:89]: in combination with my comment above, I recommend adding a test where callback is nil

File: lib/hawkular/operations/operations_api.rb (hunk 3 above)
Comment (pilhuhn) [inline lib/hawkular/operations/operations_api.rb]: While you are right @abonas the situation is even more different as the `callback.perform` works on the failure callback, but that is not checked for. 

The only situation this method should `fail` is if no error-callback is given. Otherwise the error-callback should be called with the problem (message).

Comment (rubenvp8510) [inline lib/hawkular/operations/operations_api.rb]: Just for clarify, this should fail and throw an exception only in the following cases:

```
client.invoke_generic_operation(noop) do |on|
  on.success do |_data|
    puts 'This should have failed'
  end
end
```

`client.invoke_generic_operation(noop)
`

but shouldn't fail in this case:

```
client.invoke_generic_operation(noop) do |on|
  on.success do |_data|
    puts 'This should have failed'
  end

  on.failure do |error|
    puts error
  end
end

```

Is that correct? If this is the case, I'll do the changes.

Comment (pilhuhn) [inline lib/hawkular/operations/operations_api.rb]: Yes. It should only fail if there is no error callback at all

Am 12. Juli 2016 5:17:49 nachm. schrieb Ruben Vargas 
notifications@github.com:

File: spec/integration/operations_spec.rb [hunk 5]
```diff
@@ -86,6 +86,82 @@ module Hawkular::Operations::RSpec
...
+                                      })
+
+        noop = { operationName: 'noop' }
+        expect do
+          client.invoke_generic_operation(noop) do |on|
+            on.success do |_data|
+              fail 'This should have failed'
+            end
+          end
+        end.to raise_error(ArgumentError, 'Hash property resourcePath must be specified')
```
Comment (pilhuhn) [inline spec/integration/operations_spec.rb]: This (and above) looks odd to me - not the failing per se, but rather the given reason text.
It should have raised the `ArgumentError` because there is no callback.

File: spec/integration/operations_spec.rb [hunk 6]
```diff
@@ -86,6 +86,82 @@ module Hawkular::Operations::RSpec
...
+                                        username: 'jdoe',
+                                        password: 'password'
+                                      },
+                                      options: {
+                                        tenant: 'hawkular'
+                                      })
+
+        noop = { operationName: 'noop' }
+        expect { client.invoke_generic_operation(noop) }.to raise_error(ArgumentError,
+                                                                        'Hash property resourcePath must be specified')
```
Comment (pilhuhn) [inline spec/integration/operations_spec.rb]: (see below)

File: lib/hawkular/operations/operations_api.rb [hunk 7]
```diff
@@ -215,7 +217,10 @@ def check_pre_conditions(hash = {}, params = [], &callback)
       fail 'callback must have the perform method defined. include Hawkular::Operations' unless
           callback.nil? || callback.respond_to?('perform')
       params.each do |property|
-        fail "Hash property #{property} must be specified" if hash[property].nil?
+        next unless hash[property].nil?
+        err_message = "Hash property #{property} must be specified"
+        fail(ArgumentError, err_message) if callback.nil?
```
Comment (pilhuhn) [inline lib/hawkular/operations/operations_api.rb]: See below - if all callbacks are missing or only the error callback is missing, we need to fail, saying so.
Once the error callback is present, all failures need to be returned via the error callback.
